storage-proofs/src/stacked/proof.rs

```diff
@@ -16,395 +18,603 @@ use crate::stacked::{
-                        let buf =
-                            data_at_node(&layer_labels, *parent as usize).expect("invalid node");
-                        hasher.input(buf);
-                    }
-
-                    // Expander parents
-                    // This will happen for all layers > 1
-                    if let Some(ref parents_data) = exp_parents_data {
-                        for parent in parents.iter().skip(base_parents_count) {
-                            let buf =
-                                data_at_node(parents_data, *parent as usize).expect("invalid node");
-                            hasher.input(&buf);
-                        }
-                    }
-                }
+                let key = create_key(
+                    window_graph,
+                    base_hasher.clone(),
+                    &parents,
+                    exp_parents_data.as_ref(),
+                    &layer_labels,
+                    window_index,
+                    node,
+                );
 
                 let start = data_at_node_offset(node);
                 let end = start + NODE_SIZE;
 
-                // finalize the key
-                let mut key = hasher.result();
-                // strip last two bits, to ensure result is in Fr.
-                key[31] &= 0b0011_1111;
-
                 // store the newly generated key
                 layer_labels[start..end].copy_from_slice(&key[..]);
 
-                if with_hashing {
-                    let sender_index = node / chunk_len;
-                    let sender = &cs_handle.as_ref().unwrap()[sender_index].1;
-                    if layer == 1 {
-                        // Initialize hashes on layer 1.
-                        sender
-                            .send(Message::Init(node, key))
-                            .expect("failed to init hasher");
-                    } else {
-                        // Update hashes for all other layers.
-                        sender
-                            .send(Message::Hash(node, key))
-                            .expect("failed to update hasher");
-                    }
-                }
-            }
-
-            if with_hashing && layer == layers {
-                // Finalize column hashes.
-                for (_, sender) in cs_handle.as_ref().unwrap().iter() {
-                    sender
-                        .send(Message::Done)
-                        .expect("failed to finalize hasher");
+                if layer == layers {
+                    // on the last layer we encode the data
+                    let keyd = H::Domain::try_from_bytes(&key).unwrap();
+                    let data_node = H::Domain::try_from_bytes(
+                        &data_chunk[node * NODE_SIZE..(node + 1) * NODE_SIZE],
+                    )
+                    .unwrap();
+                    let encoded_node = decode::<H::Domain>(keyd, data_node);
```

Comment:
Let's call this `decoded_node`, for my sanity.

Reply:
fixed